File: src/org/stormgears/utils/dsio/POVButton.java
package org.stormgears.utils.dsio;

import edu.wpi.first.wpilibj.Joystick;
import edu.wpi.first.wpilibj.buttons.Button;
import edu.wpi.first.wpilibj.command.Command;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

public class POVButton extends Button /* WPI button! */ implements StormButton {
	private static final Logger logger = LogManager.getLogger(POVButton.class);

	public enum Direction {
		Up(0, 315, 45),
		Right(90, 45, 135),
		Down(180, 135, 225),
		Left(270, 315, 225);

		private final int deg;
		private final int upper;
		private final int lower;

		Direction(int deg, int lower, int upper) {
			this.deg = deg;
			this.lower = lower;
			this.upper = upper;
		}

		private boolean match(int d) {
			return d == deg || d == lower || d == upper;
		}
	}

	private final Joystick joystick;
	private final Direction direction;

	public POVButton(Joystick joystick, Direction direction) {
		this.joystick = joystick;
		this.direction = direction;

		this.whenPressed(() -> logger.info("POV button {} pressed - {}", direction, joystick.getName()));
	}

	@Override
	public boolean get() {
		int d = this.joystick.getPOV();
		return d != -1 && direction.match(d);
	}

	@Override
	public void whenPressed(Runnable callback) {
		super.whenPressed(new Command() {
			@Override
			protected boolean isFinished() {
				return true;
			}

			@Override
			protected void execute() {
				callback.run();
			}
		});
	}
}

File: src/main/java/org/stormgears/utils/dsio/POVButton.kt
package org.stormgears.utils.dsio

import edu.wpi.first.wpilibj.Joystick
import edu.wpi.first.wpilibj.buttons.Button
import org.apache.logging.log4j.LogManager
import org.stormgears.utils.LambdaCommand
import org.stormgears.utils.RunnableCommand

class POVButton(private val joystick: Joystick, private val direction: Direction) : Button /* WPI button! */(), IButton {

	enum class Direction(private val deg: Int, private val lower: Int, private val upper: Int) {
		Up(0, 315, 45),
		Right(90, 45, 135),
		Down(180, 135, 225),
		Left(270, 315, 225);

		internal fun match(d: Int): Boolean {
			return d == deg || d == lower || d == upper
		}
	}

	init {
		this.whenPressed { logger.trace("POV button {} pressed - {}", direction, joystick.name) }
	}

	override fun get(): Boolean {
		val d = this.joystick.pov
		return d != -1 && direction.match(d)
	}

	override fun whenPressed(callback: Runnable) {
		super.whenPressed(RunnableCommand(callback))
	}

	override fun whenPressed(callback: () -> Unit) {
		super.whenPressed(LambdaCommand(callback))
	}

	override fun whenReleased(callback: Runnable) {
		super.whenReleased(RunnableCommand(callback))
	}

	override fun whenReleased(callback: () -> Unit) {
		super.whenReleased(LambdaCommand(callback))
	}

	companion object {
		private val logger = LogManager.getLogger(POVButton::class.java)
	}
}
